### PR TITLE
Add IntroduceFunctionFix

### PR DESCRIPTION
**Pull request by grenki**

### CODE REVIEW

File: src/com/goide/inspections/unresolved/GoIntroduceFunctionFix.java [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+    Map<String, GoImportSpec> importMap = ((GoFile)file).getImportedPackagesMap();
+    for (GoType type : types) {
+      resultTypes.add("i " + checkType(file, type, importMap));
+    }
+    return resultTypes;
+  }
+
+  @NotNull
+  private static String checkType(@NotNull PsiFile file, @Nullable GoType type, Map<String, GoImportSpec> importMap) {
+    if (type == null) return "interface{}";
```
Comment (zolotov) [inline src/com/goide/inspections/unresolved/GoIntroduceFunctionFix.java]: check when type is a chan of som imported type

File: src/com/goide/refactor/GoRefactoringUtil.java [hunk 2]
```diff
@@ -74,6 +82,84 @@ public static PsiElement findAnchor(@NotNull List<PsiElement> occurrences, @Null
     return statement == null ? GoPsiImplUtil.getTopLevelDeclaration(first) : statement;
   }
 
+  @NotNull
+  public static Expression getParameterNameSuggestedExpression(GoExpression expression) {
```
Comment (zolotov) [inline src/com/goide/refactor/GoRefactoringUtil.java]: create

File: src/com/goide/refactor/GoRefactoringUtil.java [hunk 3]
```diff
@@ -74,6 +82,84 @@ public static PsiElement findAnchor(@NotNull List<PsiElement> occurrences, @Null
...
+      Set<String> parameterNames = new HashSet<String>();
+      for (GoParameterDeclaration paramDecl : parameters.getParameterDeclarationList()) {
+        for (GoParamDefinition paramDef : paramDecl.getParamDefinitionList()) {
+          if (parameter == paramDef) continue;
+          parameterNames.add(paramDef.getName());
+        }
+      }
+
+      for (String name : myNames) {
+        if (parameterNames.contains(name)) {
```
Comment (zolotov) [inline src/com/goide/refactor/GoRefactoringUtil.java]: please investigate something about com.intellij.util.text.UniqueNameGenerator

File: src/com/goide/inspections/unresolved/GoIntroduceFunctionFix.java (hunk 1 above)
Comment (zolotov) [inline src/com/goide/inspections/unresolved/GoIntroduceFunctionFix.java]: try to reuse `com.goide.GoDocumentationProvider#getTypePresentation` with an ability to substutute type-to-string converter via function argument

**Comment by zolotov** [comment]
merged manually

**Comment by dlsniper** [comment]
@grenki / @zolotov thank you so much for this, I'm using it right now and it's so good to have it!

**Pull request closed without merging**